scripts/generate_upgrade_syncset.py

```diff
@@ -0,0 +1,301 @@
+    outer['kind'] = 'List'
+
+    items = CS()
+    for cluster_id, cluster_name in clusters.items():
+        uc = generate_upgradeconfig(start_time, version, channel)
+        ss_namespace = 'uhc-{}-{}'.format(environment, cluster_id)
+        data = CM()
+        data['apiVersion'] = SS_API_VERSION
+        data['kind'] = 'SyncSet'
+        data['metadata'] = {
+            'labels': {
+                'api.openshift.com/id': cluster_id,
+                'api.openshift.com/name': cluster_name,
+            },
+            'name': SS_NAME,
+            'namespace': ss_namespace,
+        }
+
+        data['spec'] = {
+            'clusterDeploymentRefs': [
+                {
+                    'name': dq(cluster_name),
+                }
+            ],
+            'resourceApplyMode': 'Upsert',
+            'resources': [
+                uc
+            ],
+        }
+        items.append(data)
+
+    outer['items'] = items
+
+    yaml.preserve_quotes = True
+    yaml.dump(outer, sys.stdout, Dumper=yaml.RoundTripDumper)
+
+
+def init_argparse():
+    """
+    Initialises the command-line argument parser
+    :return: parser instance
+    """
+    parser = argparse.ArgumentParser(description='UpgradeConfig SelectorSyncSet generator',
+                                     usage='%(prog)s [options]')
+
+    parser.add_argument('--subgroup', required=False, choices=['all', 'prod', 'nonprod'], default='all',
+                        help='Sub-cluster-group to target (default=all)')
+    parser.add_argument('--start-time', required=False, type=valid_date,
+                        help='Timestamp at which upgrade should commence (eg. 2020-01-25T00:00:00Z)')
+    parser.add_argument('--version', required=True, type=valid_version, help='Target upgrade version [required]')
+    parser.add_argument('--channel', required=True, help='Target upgrade channel [required]')
+    parser.add_argument('--force', '-f', required=False, default=False, help='Force upgrade [optional, default false]')
+    parser.add_argument('--environment', '-e', required=False, choices=['production', 'staging', 'integration'],
+                        default='production', help='Hive environment [optional, default production]')
+
+    cluster_group = parser.add_argument_group()
+    cluster_group.add_argument('--cluster-id', '-c', required=False, action='append', help='Cluster ID to target')
+    cluster_group.add_argument('--group', '-g', required=False, action='append',
+                               choices=['sd', 'sre', 'cicd', 'internal', 'snowflake', 'external'],
+                               help='Cluster group to target')
+
+    return parser
+
+
+def valid_version(v):
+    """
+    Verify the supplied version adheres to formatting conventions
+    :param v: version to verify
+    :return: the unmodified version if it parses successfully
+    :raises ArgumentError if version is invalid
+    """
+    p = re.compile('^4\\.[0-9]+\\.[0-9]+')
```

Comment:
Recommend using raw strings here. It allows you to get rid of the backslash doubling, but it's also best practice for regexes anyway.
```suggestion
    p = re.compile(r'^4\.[0-9]+\.[0-9]+')
```

Reply:
Be careful with a firm regex. Will this ever need to match rc builds, or nightlies? I don't think they adhere to the `x.y.z` where x, y, z are each integers.